sdk/go/pulumi/context.go

```diff
@@ -308,6 +308,182 @@ func (ctx *Context) Invoke(tok string, args interface{}, result interface{}, opt
 	return nil
 }
 
+// Call will invoke a provider call function, identified by its token tok.
+//
+// output is used to determine the output type to return; self is optional for methods.
+func (ctx *Context) Call(tok string, args Input, output Output, self Resource, opts ...InvokeOption) (Output, error) {
```

Comment:
I'd love feedback on the signature. In particular `output Output`. It is used to pass a "dummy" output, which is only used to be able to determine the type of the output to return. The actual output we return is a new one that we create.

For example:

```go
func (c *Component) GetMessage(ctx *pulumi.Context, args *ComponentGetMessageArgs) (ComponentGetMessageResultOutput, error) {
 	out, err := ctx.Call("testcomponent:index:Component/getMessage", args, ComponentGetMessageResultOutput{}, c)
 	if err != nil {
 		return ComponentGetMessageResultOutput{}, err
 	}
 	return out.(ComponentGetMessageResultOutput), nil
 }
```

Is this weird? Is there a "better" way?

Reply:
`reflect.Type` and pass `reflect.TypeOf(ComponentGetMessageResultOutput{})` comes to mind as an option.

Reply:
For `self Resource`, is only the URN/id actually used?

Reply:
> `reflect.Type` and pass `reflect.TypeOf(ComponentGetMessageResultOutput{})` comes to mind as an option.

Yep, certainly an option. It's a more primitive type and may reduce some confusion around "Why am I passing an `Output` here?" On the other hand, then they have to write `reflect.TypeOf(ComponentGetMessageResultOutput{})` when we could just do that for them. And someone may end up passing some other `reflect.Type` completely unrelated to an output.

I think I'll stick with what I've got as it forces passing an output. Users aren't typically going to be using `Call` directly anyway (instead they'll be using generated methods/functions in our SDKs).

> For `self Resource`, is only the URN/id actually used?

It'll be used as the `__self__` argument, which is marshaled as a resource reference (which ends up using the URN/id). It's also used for the provider/version options, if they're available on the resource.